Implement PlainTime

ashutoshvarma opened the pull request:
Part of #21

- [ ] PlainTime
  - [x] compare (https://github.com/ColinEberhardt/assemblyscript-temporal/pull/8)
  - [x] equals (https://github.com/ColinEberhardt/assemblyscript-temporal/pull/8)
  - [x] from
  - [x] add
  - [x] subtract
  - [x] toString
  - [x] with
  - [x] until
  - [x] since
  - [ ] round
  - [ ] toJSON
  - [ ] valueOf
  - [x] toPlainDateTime


Other fixes include in PR
- fix PlainDateTime.fromString for "1976-11-18T15:23:30.1Z" family              [[2d0abd5](https://github.com/ColinEberhardt/assemblyscript-temporal/pull/25/commits/2d0abd54b9dd64c74271f1393d87db3d4286c59a)]
- fix PlainDateTime.fromString for "1976-11-18T152330.1+00:00" family      [[73b57ec](https://github.com/ColinEberhardt/assemblyscript-temporal/pull/25/commits/73b57ec507ace84a40fc05fc6e57d70ce2a1af78)]
- fix #29 : balanceDuration() for largestUnit greater than Days                 [[5c36d22](https://github.com/ColinEberhardt/assemblyscript-temporal/pull/25/commits/5c36d2263079106350570bb89c3c9f5d07756c71)]
- fix Duration.toString() for negative values [[6cc1621](https://github.com/ColinEberhardt/assemblyscript-temporal/pull/25/commits/6cc16218dee06247993293e70fe657478de05468)]
- fix trailing zero in Duration.toString()

ashutoshvarma committed:
add tests for PlainTime constructor

ashutoshvarma committed:
implemented PlainDateTime.add, subtract, fromTimeLike

ashutoshvarma committed:
fix PlainDateTime.fromString for "1976-11-18T15:23:30.1Z" family

ashutoshvarma committed:
implement PlainDateTime.toPlainTime()

ashutoshvarma committed:
fix PlainDateTime.fromString for "1976-11-18T152330.1+00:00" family

ashutoshvarma committed:
implement PlainTime.toString(), fromPlainDateTime(), fromString(), from()

ashutoshvarma committed:
add more tests for PlainTime

- toString()
- from()
- compare()
- add()
- subtract()

ashutoshvarma commented:
From failing tests log, seems FP rounding error
```log
[Fail]: ✖ 15:23:30.123456789.add({ nanoseconds: 300 })
[Actual]: "15:23:30.12345708899999999"
[Expected]: "15:23:30.123457089"
```

How should we handle floating-point rounding errors?

MaxGraey commented on assembly/plaintime.ts:
```diff
@@ -14,33 +32,161 @@ export class PlainTime {
+      toPaddedString(this.minute) +
+      ":" +
+      toPaddedString(this.second) +
+      (this.nanosecond != 0 || this.microsecond != 0 || this.millisecond != 0
```
```suggestion
      ((this.nanosecond | this.microsecond | this.millisecond) != 0
```

MaxGraey commented on assembly/plaintime.ts:
```diff
@@ -64,4 +210,86 @@ export class PlainTime {
+    const duration =
+      durationToAdd instanceof DurationLike
+        ? durationToAdd.toDuration()
+        : // @ts-ignore TS2352
+          (durationToAdd as Duration);
```
```suggestion
        // @ts-ignore TS2352
        : (durationToAdd as Duration);
```

MaxGraey commented on assembly/plaintime.ts:
```diff
@@ -64,4 +210,86 @@ export class PlainTime {
+    const duration =
+      durationToSubtract instanceof DurationLike
+        ? durationToSubtract.toDuration()
+        : // @ts-ignore TS2352
+          (durationToSubtract as Duration);
```
```suggestion
        // @ts-ignore TS2352
        : (durationToSubtract as Duration);
```

MaxGraey commented:
> How should we handle floating-point rounding errors?

Such imprecise rarely happens in `dtoa` implementaion in AS. I have plan rewrite that to more modern Ryu's implementation. But currently we could just ignore this imprecision and add some TODO comment

ashutoshvarma committed:
ignore failing PlainTime.add test as todo

- Failing due to FP imprecision in AS
- https://github.com/ColinEberhardt/assemblyscript-temporal/pull/25#issuecomment-812995856

ashutoshvarma committed:
fix minor code formatting

Apply suggestions from code review

Co-authored-by: Max Graey <maxgraey@gmail.com>

ashutoshvarma committed:
implement PlainTime.with

ashutoshvarma committed:
implement PlainTime.toPlainDateTime()

MaxGraey commented on assembly/plaintime.ts:
```diff
@@ -64,4 +238,86 @@ export class PlainTime {
+    const duration =
+      durationToAdd instanceof DurationLike
+        ? durationToAdd.toDuration()
+        : // @ts-ignore TS2352
+          (durationToAdd as Duration);
```
```suggestion
        // @ts-ignore TS2352
        : (durationToAdd as Duration);
```

ashutoshvarma replied:
Ohh crap, I should disable eslint in vscode 😅

MaxGraey commented on assembly/plaintime.ts:
```diff
@@ -64,4 +238,86 @@ export class PlainTime {
+    const duration =
+      durationToSubtract instanceof DurationLike
+        ? durationToSubtract.toDuration()
+        : // @ts-ignore TS2352
+          (durationToSubtract as Duration);
```
```suggestion
        // @ts-ignore TS2352
        : (durationToSubtract as Duration);
```

MaxGraey commented on assembly/plaintime.ts:
```diff
@@ -14,33 +34,187 @@ export class PlainTime {
+  }
+
+  toPlainDateTime(
+    dateLike: DateLike = { year: 0, month: 0, day: 0 }
```
According to this:
https://github.com/tc39/proposal-temporal/blob/2f36d91332aeebc3a66082600c8eef57b751fbd5/polyfill/test/PlainDate/prototype/toPlainDateTime/time-undefined.js#L10

better use this approach:
```ts
toPlainDateTime(dateLike: DateLike | null = null): PlainDateTime {
  let year = 0, month = 0, day = 0;
  if (dateLike !== null) {
     year  = coalesce(dateLike!.year, 0);
     month = coalesce(dateLike!.month, 0);
     day   = coalesce(dateLike!.day, 0);
  }
  return new PlainDateTime(
    year,
    month,
    day,
    this.hour,
    this.minute,
    this.second,
    this.millisecond,
    this.microsecond,
    this.nanosecond
  );
}
```

ashutoshvarma replied:
Done 👍

ashutoshvarma committed:
update PlainTime.toPlainDateTime to handle undefined properly

Co-authored-by: Max Graey <maxgraey@gmail.com>

MaxGraey commented on assembly/plaintime.ts:
```diff
@@ -64,4 +242,86 @@ export class PlainTime {
     return ord(a.nanosecond, b.nanosecond);
   }
+
+  add<T>(durationToAdd: T): PlainTime {
```
```suggestion
  add<T = DurationLike>(durationToAdd: T): PlainTime {
```

MaxGraey replied:
And after this you could remove all type hints in tests like 
`.add<DurationLike>({ hours: 2, ... })` to `.add({ hours: 2, ... })`

ashutoshvarma replied:
Done

MaxGraey commented on assembly/plaintime.ts:
```diff
@@ -64,4 +242,86 @@ export class PlainTime {
+    );
+  }
+
+  subtract<T>(durationToSubtract: T): PlainTime {
```
```suggestion
  subtract<T = DurationLike>(durationToSubtract: T): PlainTime {
```

ashutoshvarma replied:
Done

ashutoshvarma committed:
add default type in PlainTime.add() and subtract()

ashutoshvarma committed:
add regulateTime()

- rejectTime() yet to be implemented

ashutoshvarma committed:
add differenceTime() in utils

ashutoshvarma committed:
improve PlainTime.add() and subtract

make them more close to tc39 implementation

ashutoshvarma commented:
@MaxGraey shouldn't it be 8 instead of 4 in `sign` ? or am I missing something?
https://github.com/ColinEberhardt/assemblyscript-temporal/blob/1d02f91541d1090e8bbf9650b6f5a793f91a84b7/assembly/utils.ts#L177-L181

ashutoshvarma committed:
implement PlainTime.until()

MaxGraey commented:
right! Good catch!

ashutoshvarma commented:
Also while working on `util()`, I get to know that `balanceDuration` is broken when `largestUnit` is greater than `TimeComponent.days` .

I will create an issue for it shortly

ashutoshvarma commented:
> right! Good catch!

Should I add a commit for it in this PR ? `PlainTime.until()` tests are failing because of it too

MaxGraey commented:
I already made PR with fix: https://github.com/ColinEberhardt/assemblyscript-temporal/pull/28

ashutoshvarma committed:
fix #29 : balanceDuration() for largestUnit greater than Days

MaxGraey commented on assembly/utils.ts:
```diff
@@ -395,83 +395,87 @@ export function balanceDuration(
-  seconds = 0;
-  minutes = 0;
-  hours = 0;
+  const sig = sign(nanosecondsI64);
```
```suggestion
  const sig = i32(sign(nanosecondsI64));
```

ashutoshvarma replied:
Does casting sig to `i32` change something?

MaxGraey replied:
See next suggestion. Main idea is avoiding 64-bit arithmetic as much as possible due to some wasm engines or some 32-bit platforms may be significantly slower with 64-bit arithmetic

ashutoshvarma replied:
Ohh, I see.

MaxGraey commented on assembly/utils.ts:
```diff
@@ -482,13 +486,13 @@ export function balanceDuration(
-    milliseconds * sig,
-    microseconds * sig,
-    nanoseconds * sig
+    i32(daysI64),
+    i32(hoursI64 * sig),
+    i32(minutesI64 * sig),
+    i32(secondsI64 * sig),
+    i32(millisecondsI64 * sig),
+    i32(microsecondsI64 * sig),
+    i32(nanosecondsI64 * sig)
```
```suggestion
    i32(daysI64),
    i32(hoursI64) * sig,
    i32(minutesI64) * sig,
    i32(secondsI64) * sig,
    i32(millisecondsI64) * sig,
    i32(microsecondsI64) * sig,
    i32(nanosecondsI64) * sig
```

ashutoshvarma committed:
balanceDuration: convert sig to i32 from i64

see https://github.com/ColinEberhardt/assemblyscript-temporal/pull/25/files#r607097375
Apply suggestions from code review

Co-authored-by: Max Graey <maxgraey@gmail.com>

ashutoshvarma committed:
fix Duration.toString() for negative values

ashutoshvarma commented:
https://github.com/ColinEberhardt/assemblyscript-temporal/blob/1d02f91541d1090e8bbf9650b6f5a793f91a84b7/assembly/duration.ts#L71-L78

@MaxGraey is there a way to remove the trailing `.0` from the f64 string representation? I tried reparsing string back using `parseFloat` and then `toString` but it didn't work

From CI logs https://github.com/ColinEberhardt/assemblyscript-temporal/runs/2270205795
```
[Fail]: ✖ the default largest unit is at least hours
[Actual]: "PT6H52M42.0S"
[Expected]: "PT6H52M42S"
```

MaxGraey commented:
> is there a way to remove the trailing .0 from the f64 string representation? I tried reparsing string back using 

Trailing zero for serialized f64/f32 strings is a deliberate decision and convention of AssemblyScript which differ from JS/TS. So I suggest slightly modify original tests and add `.0` when this required

ashutoshvarma committed:
update PlainTime.untill tests

ashutoshvarma committed:
implement PlainTime.since

ashutoshvarma commented:
I think this PR is good to go now!

MaxGraey commented:
> > is there a way to remove the trailing .0 from the f64 string representation? I tried reparsing string back using
> 
> Trailing zero for serialized f64/f32 strings is a deliberate decision and convention of AssemblyScript which differ from JS/TS. So I suggest slightly modify original tests and add `.0` when this required

Another option:
```ts
function stringify(value: f64): string {
   return F64.isSafeInteger(value) ? i64(value).toString() : value.toString();
}
```

ashutoshvarma committed:
Revert "update PlainTime.untill tests"

This reverts commit 65b683792523b002ae8ac879a5394e2b75cbdb32.

ashutoshvarma committed:
fix trailing zero in Duration.toString()

ashutoshvarma commented:
> ```ts
> function stringify(value: f64): string {
>    return F64.isSafeInteger(value) ? i64(value).toString() : value.toString();
> }
> ```

This is great! With this `Duration.toString(`) will be more close to tc39 implementation

ColinEberhardt committed:
Merge branch 'master' into master

ColinEberhardt commented:
> I think this PR is good to go now!

Thank you - great work 👍
